text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1563 @@
+
+> [..] we don't think there's much use of this in the wild,
+> and we have an automated fix.
+
+Thus, we believe it is reasonable to expedite the schedule.
+
+# Prior art
+[prior-art]: #prior-art
+
+## Haskell
+
+In Haskell it possible to type ascribe an expression like so
+(here using the REPL `ghci`):
+
+```haskell
+ghci> 1 + 1 :: Int -- Type ascribing 1 + 1 to the type Int.
+2
+
+ghci> 1 + 1 :: Bool -- And to Bool, which is wrong.
+
+<interactive>:4:1: error:
+    • No instance for (Num Bool) arising from a use of ‘+’
+    • In the expression: 1 + 1 :: Bool
+      In an equation for ‘it’: it = 1 + 1 :: Bool
+```
+
+It should be noted that Haskell, just like Rust, allows a user to apply types
+to a polymorphic function explicitly:
+
+```haskell
+{-# LANGUAGE TypeApplications #-}
+
+id :: forall a. a -> a
+id x = x
+
+foo = id @Int 1 -- We apply Int to the type variable 'a' above.
+```
+
+This would correspond roughly to:
+
+```rust
+fn id<T>(x: T) -> T { x }
+
+let foo = id::<i32>(1);
+```
+
+Note in particular here that the Haskell version uses the same token for
+annotating the function signature and for ascribing types on expressions.
+
+As with this RFC, you can also type ascribe inside patterns in Haskell:
+
+```haskell
+ghci> :set -XScopedTypeVariables
+ghci> foo (x :: Int, y :: Bool) = if y then x + 1 else x - 1
+ghci> :t foo
+foo :: (Int, Bool) -> Int
+```
+
+## PureScript
+
+[PureScript]: https://github.com/purescript/documentation/blob/master/language/Types.md#type-annotations
+
+Being a dialect of Haskell, [PureScript] also allow users to ascribe expressions.
+
+## Idris
+
+Idris annotates its function definitions like so:
+
+```idris
+id : a -> a
+id x = x
+```
+
+However, Idris does not have a built-in mechanism to type-ascribe expressions.
+Instead, you use the library defined function `the`:
+
+```idris
+the : (a : Type) -> (value : a) -> a
+the _ = id
+```
+
+You may then write `the Nat x` for the equivalent of `x : Nat`.
+
+## Scala
+
+[scala_annot]: https://docs.scala-lang.org/style/types.html#annotations
+[scala_ascribe]: https://docs.scala-lang.org/style/types.html#ascription
+
+Scala supports both what it calls ["type annotations"][scala_annot] and
+["type ascription"][scala_ascribe].
+ 
+For example, you may write (type annotation):
+
+```scala
+val s = "Alan": String
+```
+
+You may also write (type ascription, upcasting):
+
+```scala
+val s = s: Object
+```
+
+Note in particular that Scala does take sub-typing (of a different kind) into
+account in this syntax.
+
+## F*
+
+[fstar]: https://www.fstar-lang.org/
+[fstar_ascribe]: https://github.com/FStarLang/FStar/wiki/F*-symbols-reference
+
+[F*][fstar] allows users to [type ascribe][fstar_ascribe] using the symbol `<:`.
+For example:
+
+```ocaml
+module Ascribe
+
+val x : string
+let x = "foo" <: string
+```
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+There are none.
+
+# Possible future work
+[possible future work]: #possible-future-work
+
```

Comment:
Not sure whether you want to think about it now, or leave it for the future, but I'd love to be able to use `impl Trait` in type ascription. I'm not sure what the semantics would be though (I'd like `e: impl T` to mean that the type of `e` implements `T`, but I'm not sure how that would interact with existing uses in function args and return types).

Reply:
Good catch; I think I can add it now.
I think the semantics should be rather well defined:

1. In `fn foo(Wrapping(x: impl Trait)) { .. }` it acts just like `universal_impl_trait` does otherwise.

2. In the function body it acts like `let x: impl Trait = ..;` does. The semantics of this is defined by https://github.com/rust-lang/rfcs/pull/2071.